File: y22/m07/d25/target_number.py
# 다시 풀어본 풀이

same_count = 0


def solution(numbers, target):
    global same_count

    def dfs(numbers, target, index):
        global same_count
        if index == len(numbers) - 1:
            if sum(numbers) == target:
                same_count += 1
            return

        dfs(numbers, target, index + 1)
        numbers[index + 1] *= -1
        dfs(numbers, target, index + 1)

    dfs(numbers, target, -1)

    return same_count


# DFS 풀이
def solution(numbers, target):
    def dfs(numbers, target, index):
        same_count = 0
        if index == len(numbers):
            if sum(numbers) == target:
                return 1
            return 0

        same_count += dfs(numbers, target, index + 1)
        numbers[index] *= -1
        same_count += dfs(numbers, target, index + 1)
        return same_count

    same_count = dfs(numbers, target, 0)

    return same_count


# BFS 풀이
def solution(numbers, target):
    answer = 0
    q = []
    n = len(numbers)
    q.append((numbers[0], 0))
    q.append((-1 * numbers[0], 0))
    while q:
        temp, idx = q.pop()
        idx += 1
        if idx < n:
            q.append((temp + numbers[idx], idx))
            q.append((temp - numbers[idx], idx))
        else:
            if temp == target:
                answer += 1
    return answer

File: y22/m09/d07/TargetNumber.java
package y22.m09.d07;

public class TargetNumber {
    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] numbers = new int[]{1, 1, 1, 1, 1};
        int target = 3;
        int correctAnswer = 5;
        int answer = solution.solution(numbers, target);

        System.out.println(answer == correctAnswer);
        System.out.println("answer = " + answer);
        System.out.println("correctAnswer = " + correctAnswer);
    }
}

class Solution {
    int answer = 0;
    int[] numbers;
    int target;

    public int solution(int[] numbers, int target) {
        this.numbers = numbers;
        this.target = target;

        dfs(0, 0);
        return answer;
    }

    public void dfs(int index, int sum) {
        // 탈출 조건
        if (numbers.length == index) {
            if (sum == target) {
                answer++;
            }
            return;
        }

        // 수행 동작
        dfs(index + 1, sum + numbers[index]);
        dfs(index + 1, sum - numbers[index]);
    }
}

